libpromises/promises.c

```diff
@@ -170,15 +172,69 @@ Promise *DeRefCopyPromise(EvalContext *ctx, const Promise *pp)
                 {
-                    Constraint *scp = SeqAt(bp->conlist, k);
+                    // remember we reversed the Seq from what
+                    // EvalContextResolveBodyExpression() created
+                    Rval *called_rval = SeqAt(bodies_and_args, body_index);
+                    const Body *current_body = SeqAt(bodies_and_args, body_index+1);
+                    JsonElement *arg_rewrite = JsonObjectCreate(2);
+
+                    if (NULL == called_rval)
+                    {
+                        // nothing needed, this is not an inherit_from rval
+                    }
+                    else if (RVAL_TYPE_SCALAR == called_rval->type)
+                    {
+                        // We leave the parameters as they were.
 
-                    if (IsDefinedClass(ctx, scp->classes))
+                        // Unless the current body matches the
+                        // parameters of the inherited body, there
+                        // will be unexpanded variables. But the
+                        // alternative is to match up body and fncall
+                        // arguments, which is not trivial.
+                    }
+                    else if (RVAL_TYPE_FNCALL == called_rval->type)
                     {
-                        Rval returnval = ExpandPrivateRval(ctx, NULL, "body", scp->rval.item, scp->rval.type);
-                        PromiseAppendConstraint(pcopy, scp->lval, returnval, false);
+                        const Rlist *call_args = NULL;
+                        const Rlist *body_args = NULL;
+                        // step through the body and call args
+                        for (call_args = RvalFnCallValue(*called_rval)->args, body_args = current_body->args;
+                             call_args && body_args;
+                             call_args = call_args->next, body_args = body_args->next)
+                        {
+                            JsonObjectAppendString(arg_rewrite, RlistScalarValue(body_args), RlistScalarValue(call_args));
+                        }
                     }
+
+                    for (size_t k = 0; k < SeqLength(current_body->conlist); k++)
+                    {
+                        Constraint *scp = SeqAt(current_body->conlist, k);
+
+                        // we don't copy the inherit_from attribute or associated call
+                        if (strcmp("inherit_from", scp->lval) == 0)
+                        {
+                            continue;
+                        }
+
+                        if (IsDefinedClass(ctx, scp->classes))
+                        {
+                            Rval returnval = ExpandPrivateRval(ctx, NULL, "body", scp->rval.item, scp->rval.type);
+
+                            Rval rewrite = RvalCopyRewriter(returnval, arg_rewrite);
+                            RvalDestroy(returnval);
+                            returnval = ExpandPrivateRval(ctx, NULL, "body", rewrite.item, rewrite.type);
```

Comment:
The bug is here: we rewrite the body parameters twice to get inherited parameters expanded.  Working on it.